### PR TITLE
feat(kubectl plugin): add get block-devices command to list all/usable disks per node

### PR DESCRIPTION
**Pull request by Abhinandan-Purkait**
Usage/Output All, Usable

```
➜  mayastor-control-plane (device-inv) kubectl mayastor get block-devices <node-name> --all

 DEVNAME          DEVTYPE    SIZE       AVAILABLE  MODEL                       DEVPATH                                                         FSTYPE  FSUUID  MOUNTPOINT  PARTTYPE                              MAJOR            MINOR                                     DEVLINKS 
 /dev/nvme0n1     disk       167772160  no         Amazon Elastic Block Store  /devices/pci0000:00/0000:00:04.0/nvme/nvme0/nvme0n1             259     0                                                                                                                    "/dev/disk/by-path/pci-0000:00:04.0-nvme-1", "/dev/disk/by-id/nvme-nvme.1d0f-766f6c3038346136373931663365363866613164-416d617a6f6e20456c617374696320426c6f636b2053746f7265-00000001", "/dev/disk/by-id/nvme-Amazon_Elastic_Block_Store_vol084a6791f3e68fa1d" 
 /dev/nvme0n1p1   partition  167544799  no         Amazon Elastic Block Store  /devices/pci0000:00/0000:00:04.0/nvme/nvme0/nvme0n1/nvme0n1p1   259     1       ext4        4211f95f-755b-4c22-9740-17b12a609123  cloudimg-rootfs  gpt:0fc63daf-8483-4772-8e79-3d69d8477de4  "/dev/disk/by-uuid/4211f95f-755b-4c22-9740-17b12a609123", "/dev/disk/by-id/nvme-nvme.1d0f-766f6c3038346136373931663365363866613164-416d617a6f6e20456c617374696320426c6f636b2053746f7265-00000001-part1", "/dev/disk/by-partuuid/54ecbbe1-f829-4864-bd41-588e2ccc038b", "/dev/disk/by-label/cloudimg-rootfs", "/dev/disk/by-id/nvme-Amazon_Elastic_Block_Store_vol084a6791f3e68fa1d-part1", "/dev/disk/by-path/pci-0000:00:04.0-nvme-1-part1" 

```

```
➜  mayastor-control-plane (device-inv) kubectl mayastor get block-devices <node name>

 DEVNAME       DEVTYPE  SIZE       AVAILABLE  MODEL                       DEVPATH                                              MAJOR  MINOR  DEVLINKS 
 /dev/nvme3n2  disk     104857600  yes        Amazon Elastic Block Store  /devices/pci0000:00/0000:00:1e.0/nvme/nvme2/nvme3n2  259    7      "/dev/disk/by-path/pci-0000:00:1e.0-nvme-1", "/dev/disk/by-id/nvme-Amazon_Elastic_Block_Store_vol019d8523478a2a7cc", "/dev/disk/by-id/nvme-nvme.1d0f-766f6c3031396438353233343738613261376363-416d617a6f6e20456c617374696320426c6f636b2053746f7265-00000001" 

```

Signed-off-by: Abhinandan Purkait <abhinandan.purkait@mayadata.io>

### CODE REVIEW

**Review by paulyoong — APPROVED** [review]

File: control-plane/plugin/src/operations.rs [hunk 1]
```diff
@@ -42,3 +42,11 @@ pub trait ReplicaTopology {
     type ID;
     async fn topology(id: &Self::ID, output: &utils::OutputFormat);
 }
+
+/// GetBlockDevices trait.
+/// To be implemented by resources which support the 'get block-devices' operation
+#[async_trait(?Send)]
+pub trait GetBlockDevices {
```
Comment (tiagolobocastro) [inline control-plane/plugin/src/operations.rs:49]: Do we need a command to get block devices for all nodes? Which would simply get all nodes, loop and for each issue get_blockdevices rest call?

File: k8s/plugin/src/main.rs [hunk 2]
```diff
@@ -82,6 +82,14 @@ async fn execute(cli_args: CliArgs) {
...
             GetResources::Nodes => node::Nodes::list(&cli_args.output).await,
             GetResources::Node { id } => node::Node::get(&id, &cli_args.output).await,
+            GetResources::BlockDevices(bdargs) => {
+                blockdevice::BlockDevice::get_blockdevices(
+                    &bdargs.node_id(),
+                    &bdargs.all(),
+                    &cli_args.output,
+                )
+                .await
+            }
```
Comment (tiagolobocastro) [inline k8s/plugin/src/main.rs:92]: Could we add examples to the README please?

**Comment by tiagolobocastro** [comment]
Not related to these code changes but, out of curiosity, on your PR description example, after we create the pool, is the device still reported as `AVAILABLE` ?

File: control-plane/plugin/src/operations.rs (hunk 1 above)
Comment (Abhinandan-Purkait) [reply control-plane/plugin/src/operations.rs:49]: Should I add cluster wide as well? @tiagolobocastro @paulyoong

**Comment by Abhinandan-Purkait** [comment]
> Not related to these code changes but, out of curiosity, on your PR description example, after we create the pool, is the device still reported as `AVAILABLE` ?

Yes, it does. Seems like the blockdevice api, doesn't mark the `disks` used by `mayastor` pool as `not available`. The identification factor a disk to be used is not present. Need to come up with some approach to handle this.

**Comment by tiagolobocastro** [comment]
> 

Since this is the case, there's not much we could do...
we have a few options to think about:

IF we can tell the device IS in one of the `Pool`'s reported by the OpenApi, in such case we simply mark it as `NOT AVAILABLE` - we may even show the pool name as one of the properties?
But how do we know that pool is our pool? Should mayastor inspect each `AVAILABLE` device and check if it's a pool?
In any case this seems like a separate task so I propose that we simply document this on the help section of the plugin. And instead of `yes`, maybe it should read `yes ?` :) 
Thoughts?

File: control-plane/plugin/src/resources/blockdevice.rs [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+                .expect("Expected to be initialised once");
+            Ok(())
+        }
+        Err(e) => Err(e),
+    }
+}
+
+fn check_available(possible_values: Vec<String>) -> bool {
+    for value in possible_values {
+        if USED_DISKS.get().unwrap().contains(&value) {
```
Comment (tiagolobocastro) [inline control-plane/plugin/src/resources/blockdevice.rs:205]: This won't work if the diskpool is created with URI's?

File: control-plane/plugin/src/resources/blockdevice.rs [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+    },
+    rest_wrapper::RestClient,
+};
+use async_trait::async_trait;
+use once_cell::sync::OnceCell;
+use openapi::{clients::tower::Error, models::RestJsonError};
+use prettytable::Row;
+use serde::Serialize;
+
+static USED_DISKS: OnceCell<Vec<String>> = OnceCell::new();
```
Comment (tiagolobocastro) [inline control-plane/plugin/src/resources/blockdevice.rs:18]: Why do we need this? Can we not add the list of pools on the node to the new-types which we have below?

File: control-plane/plugin/src/operations.rs (hunk 1 above)
Comment (Abhinandan-Purkait) [reply control-plane/plugin/src/operations.rs:49]: Resolving for now.

File: control-plane/plugin/src/resources/blockdevice.rs (hunk 3 above)
Comment (Abhinandan-Purkait) [reply control-plane/plugin/src/resources/blockdevice.rs]: Yes, right. Thanks for pointing out. Added normalization.

File: control-plane/plugin/src/resources/blockdevice.rs (hunk 4 above)
Comment (Abhinandan-Purkait) [reply control-plane/plugin/src/resources/blockdevice.rs]: Yes, we don't need this. Thanks for pointing out. Will refactor it,

File: control-plane/plugin/src/resources/blockdevice.rs [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        let mut possible_values = vec![bd.devname.clone()];
+        for devlink in &bd.devlinks {
+            possible_values.push(devlink.clone());
+        }
+        if bd.available && check_available(used_disks, possible_values) {
+            bd.available = true;
+            transformed_bds.push(bd);
+        } else if all {
+            bd.available = false;
+            transformed_bds.push(bd);
```
Comment (tiagolobocastro) [inline control-plane/plugin/src/resources/blockdevice.rs:194]: We don't need a new vector here right? we can return blockdevices if we use iter_mut on the for loop?

File: control-plane/plugin/src/resources/mod.rs [hunk 6]
```diff
@@ -25,6 +28,8 @@ pub enum GetResources {
     Nodes,
     /// Get node with the given ID.
     Node { id: NodeId },
+    /// Get BlockDevices present on the Node. Lists usable devices by default.
```
Comment (tiagolobocastro) [inline control-plane/plugin/src/resources/mod.rs:31]: I think it's worth here mentioning here that the AVAILABLE field is not correct if the pool is not currently part of the control plane?

**Review by tiagolobocastro — APPROVED** [review]

File: control-plane/plugin/src/resources/mod.rs (hunk 6 above)
Comment (Abhinandan-Purkait) [reply control-plane/plugin/src/resources/mod.rs:31]: I have added that in the args part. As an description to the --all flag. Do we need it here as well?

File: control-plane/plugin/src/resources/blockdevice.rs (hunk 5 above)
Comment (Abhinandan-Purkait) [reply control-plane/plugin/src/resources/blockdevice.rs:194]: Yeah.

File: control-plane/plugin/src/resources/mod.rs (hunk 6 above)
Comment (tiagolobocastro) [reply control-plane/plugin/src/resources/mod.rs:31]: I'd say move it from all to here - no sure though, up to you.

Comment (Abhinandan-Purkait) [reply control-plane/plugin/src/resources/mod.rs:31]: Okay.

**Comment by Abhinandan-Purkait** [comment]
bors merge

**Pull request merged**